Optimize memory allocation for insertion

k-takata opened the pull request:
This is an additional optimization for #15588.
Reduce the memory allocation in ins_str().

If this PR is combined with #15588, `:normal 10000ix` becomes O(n):
```
$ time ./vim --clean -c 'normal 160000ix' -cq!
real    0m0.684s
user    0m0.109s
sys     0m0.108s

$ time ./vim --clean -c 'normal 1280000ix' -cq!
real    0m0.969s
user    0m0.827s
sys     0m0.077s
```

k-takata committed:
Optimize memory allocation for insertion

This is an additional optimization for #15588.
Reduce the memory allocation in ins_str().

If this PR is combined with #15588, `:normal 10000ix` becomes O(n):
```
$ time ./vim --clean -c 'normal 160000ix' -cq!
real    0m0.684s
user    0m0.109s
sys     0m0.108s

$ time ./vim --clean -c 'normal 1280000ix' -cq!
real    0m0.969s
user    0m0.827s
sys     0m0.077s
```

chrisbra commented on src/memline.c:
```diff
@@ -4236,6 +4237,7 @@ ml_flush_line(buf_T *buf)
 	    }
 	}
 	vim_free(new_line);
+	buf->b_ml.ml_line_size = 0;
```
don't you ned this a few lines below, when freeing `buf->b_ml.ml_line_ptr`?

k-takata replied:
I don't think it is necessary, but it should be safer.
Also clearing it in textprop.c might be better.

chrisbra commented on src/change.c:
```diff
@@ -1194,14 +1197,32 @@ ins_str(char_u *s)
+	newp = alloc(new_size);
+	if (newp == NULL)
+	    return;
+	curbuf->b_ml.ml_line_size = new_size;
```
That should potentially be encapsulated into a new memline()  function `ml_replace_len_size()` which takes the new size parameter and stores the new size there.

k-takata replied:
I tried to implement this, but it didn't work...

k-takata committed:
Apply review comments

Clear ml_line_size when ml_line_ptr is freed.
Add a new function to set ml_line_size.

k-takata committed:
Clear ml_line_size also in textprop.c

k-takata committed:
Remove an unnecessary static variable

k-takata committed:
Clear ml_line_size

k-takata committed:
Revert "Remove an unnecessary static variable"

This reverts commit 831fc32165551fc7cc7fd3e42326097daaf51716.

k-takata committed:
Clear ml_line_size

k-takata committed:
Revert "Apply review comments"

This reverts commit d3d05db9d5af21d5872def18d10f04b0b0bc278e.

k-takata committed:
Clear ml_line_size

k-takata committed:
Reapply "Remove an unnecessary static variable"

This reverts commit 3ada78192b5508611b9f0ff7c2c15fa04496be29.

k-takata committed:
Clear ml_line_size

k-takata committed:
Revert "Reapply "Remove an unnecessary static variable""

This reverts commit 5479858d3e7a92cbb2074775d113895032b89406.
